File: tests/html/typed_arrays_dataview_test.dart
// Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

library typed_arrays_dataview_test;

import 'package:unittest/unittest.dart';
import 'package:unittest/html_config.dart';
import 'dart:html';
import 'dart:typed_data';

main() {
  useHtmlConfiguration();

  // Only perform tests if ArrayBuffer is supported.
  if (!Platform.supportsTypedData) {
    return;
  }

  test('create', () {
    var bd = new ByteData(100);
    expect(bd.lengthInBytes, 100);
    expect(bd.offsetInBytes, 0);

    var a1 = new Uint8List.fromList([1, 2, 3, 4, 5, 6, 7, 8]);

    var bd2 = new ByteData.view(a1.buffer);
    expect(bd2.lengthInBytes, 8);
    expect(bd2.offsetInBytes, 0);

    var bd3 = new ByteData.view(a1.buffer, 2);
    expect(bd3.lengthInBytes, 6);
    expect(bd3.offsetInBytes, 2);

    var bd4 = new ByteData.view(a1.buffer, 3, 4);
    expect(bd4.lengthInBytes, 4);
    expect(bd4.offsetInBytes, 3);
  });

  test('access8', () {
    var a1 = new Uint8List.fromList([0, 0, 3, 255, 0, 0, 0, 0, 0, 0]);

    var bd = new ByteData.view(a1.buffer, 2, 6);

    expect(bd.getInt8(0), equals(3));
    expect(bd.getInt8(1), equals(-1));
    expect(bd.getUint8(0), equals(3));
    expect(bd.getUint8(1), equals(255));

    bd.setInt8(2, -56);
    expect(bd.getInt8(2), equals(-56));
    expect(bd.getUint8(2), equals(200));

    bd.setUint8(3, 200);
    expect(bd.getInt8(3), equals(-56));
    expect(bd.getUint8(3), equals(200));
  });

  test('access16', () {
    var a1 = new Uint8List.fromList([0, 0, 3, 255, 0, 0, 0, 0, 0, 0]);

    var bd = new ByteData.view(a1.buffer, 2);

    expect(bd.lengthInBytes, equals(10 - 2));

    expect(bd.getInt16(0), equals(1023));
    expect(bd.getInt16(0, Endianness.BIG_ENDIAN), equals(1023));
    expect(bd.getInt16(0, Endianness.LITTLE_ENDIAN), equals(-253));

    expect(bd.getUint16(0), equals(1023));
    expect(bd.getUint16(0, Endianness.BIG_ENDIAN), equals(1023));
    expect(bd.getUint16(0, Endianness.LITTLE_ENDIAN), equals(0xFF03));

    bd.setInt16(2, -1);
    expect(bd.getInt16(2), equals(-1));
    expect(bd.getUint16(2), equals(0xFFFF));
  });

  test('access32', () {
    var a1 = new Uint8List.fromList([0, 0, 3, 255, 0, 0, 0, 0, 0, 0]);

    var bd = new ByteData.view(a1.buffer);

    expect(bd.getInt32(0), equals(1023));
    expect(bd.getInt32(0, Endianness.BIG_ENDIAN), equals(1023));
    expect(bd.getInt32(0, Endianness.LITTLE_ENDIAN), equals(-0xFD0000));

    expect(bd.getUint32(0), equals(1023));
    expect(bd.getUint32(0, Endianness.BIG_ENDIAN), equals(1023));
    expect(bd.getUint32(0, Endianness.LITTLE_ENDIAN), equals(0xFF030000));
  });
}

File: pkg/dev_compiler/test/codegen_expected/lib/html/typed_arrays_dataview_test.js
dart_library.library('lib/html/typed_arrays_dataview_test', null, /* Imports */[
  'dart_sdk',
  'unittest'
], function load__typed_arrays_dataview_test(exports, dart_sdk, unittest) {
  'use strict';
  const core = dart_sdk.core;
  const html = dart_sdk.html;
  const typed_data = dart_sdk.typed_data;
  const _interceptors = dart_sdk._interceptors;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const html_config = unittest.html_config;
  const unittest$ = unittest.unittest;
  const src__matcher__expect = unittest.src__matcher__expect;
  const src__matcher__core_matchers = unittest.src__matcher__core_matchers;
  const typed_arrays_dataview_test = Object.create(null);
  let JSArrayOfint = () => (JSArrayOfint = dart.constFn(_interceptors.JSArray$(core.int)))();
  let VoidTodynamic = () => (VoidTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [])))();
  typed_arrays_dataview_test.main = function() {
    html_config.useHtmlConfiguration();
    if (!dart.test(html.Platform.supportsTypedData)) {
      return;
    }
    unittest$.test('create', dart.fn(() => {
      let bd = typed_data.ByteData.new(100);
      src__matcher__expect.expect(bd[dartx.lengthInBytes], 100);
      src__matcher__expect.expect(bd[dartx.offsetInBytes], 0);
      let a1 = typed_data.Uint8List.fromList(JSArrayOfint().of([1, 2, 3, 4, 5, 6, 7, 8]));
      let bd2 = typed_data.ByteData.view(a1[dartx.buffer]);
      src__matcher__expect.expect(bd2[dartx.lengthInBytes], 8);
      src__matcher__expect.expect(bd2[dartx.offsetInBytes], 0);
      let bd3 = typed_data.ByteData.view(a1[dartx.buffer], 2);
      src__matcher__expect.expect(bd3[dartx.lengthInBytes], 6);
      src__matcher__expect.expect(bd3[dartx.offsetInBytes], 2);
      let bd4 = typed_data.ByteData.view(a1[dartx.buffer], 3, 4);
      src__matcher__expect.expect(bd4[dartx.lengthInBytes], 4);
      src__matcher__expect.expect(bd4[dartx.offsetInBytes], 3);
    }, VoidTodynamic()));
    unittest$.test('access8', dart.fn(() => {
      let a1 = typed_data.Uint8List.fromList(JSArrayOfint().of([0, 0, 3, 255, 0, 0, 0, 0, 0, 0]));
      let bd = typed_data.ByteData.view(a1[dartx.buffer], 2, 6);
      src__matcher__expect.expect(bd[dartx.getInt8](0), src__matcher__core_matchers.equals(3));
      src__matcher__expect.expect(bd[dartx.getInt8](1), src__matcher__core_matchers.equals(-1));
      src__matcher__expect.expect(bd[dartx.getUint8](0), src__matcher__core_matchers.equals(3));
      src__matcher__expect.expect(bd[dartx.getUint8](1), src__matcher__core_matchers.equals(255));
      bd[dartx.setInt8](2, -56);
      src__matcher__expect.expect(bd[dartx.getInt8](2), src__matcher__core_matchers.equals(-56));
      src__matcher__expect.expect(bd[dartx.getUint8](2), src__matcher__core_matchers.equals(200));
      bd[dartx.setUint8](3, 200);
      src__matcher__expect.expect(bd[dartx.getInt8](3), src__matcher__core_matchers.equals(-56));
      src__matcher__expect.expect(bd[dartx.getUint8](3), src__matcher__core_matchers.equals(200));
    }, VoidTodynamic()));
    unittest$.test('access16', dart.fn(() => {
      let a1 = typed_data.Uint8List.fromList(JSArrayOfint().of([0, 0, 3, 255, 0, 0, 0, 0, 0, 0]));
      let bd = typed_data.ByteData.view(a1[dartx.buffer], 2);
      src__matcher__expect.expect(bd[dartx.lengthInBytes], src__matcher__core_matchers.equals(10 - 2));
      src__matcher__expect.expect(bd[dartx.getInt16](0), src__matcher__core_matchers.equals(1023));
      src__matcher__expect.expect(bd[dartx.getInt16](0, typed_data.Endianness.BIG_ENDIAN), src__matcher__core_matchers.equals(1023));
      src__matcher__expect.expect(bd[dartx.getInt16](0, typed_data.Endianness.LITTLE_ENDIAN), src__matcher__core_matchers.equals(-253));
      src__matcher__expect.expect(bd[dartx.getUint16](0), src__matcher__core_matchers.equals(1023));
      src__matcher__expect.expect(bd[dartx.getUint16](0, typed_data.Endianness.BIG_ENDIAN), src__matcher__core_matchers.equals(1023));
      src__matcher__expect.expect(bd[dartx.getUint16](0, typed_data.Endianness.LITTLE_ENDIAN), src__matcher__core_matchers.equals(65283));
      bd[dartx.setInt16](2, -1);
      src__matcher__expect.expect(bd[dartx.getInt16](2), src__matcher__core_matchers.equals(-1));
      src__matcher__expect.expect(bd[dartx.getUint16](2), src__matcher__core_matchers.equals(65535));
    }, VoidTodynamic()));
    unittest$.test('access32', dart.fn(() => {
      let a1 = typed_data.Uint8List.fromList(JSArrayOfint().of([0, 0, 3, 255, 0, 0, 0, 0, 0, 0]));
      let bd = typed_data.ByteData.view(a1[dartx.buffer]);
      src__matcher__expect.expect(bd[dartx.getInt32](0), src__matcher__core_matchers.equals(1023));
      src__matcher__expect.expect(bd[dartx.getInt32](0, typed_data.Endianness.BIG_ENDIAN), src__matcher__core_matchers.equals(1023));
      src__matcher__expect.expect(bd[dartx.getInt32](0, typed_data.Endianness.LITTLE_ENDIAN), src__matcher__core_matchers.equals(-16580608));
      src__matcher__expect.expect(bd[dartx.getUint32](0), src__matcher__core_matchers.equals(1023));
      src__matcher__expect.expect(bd[dartx.getUint32](0, typed_data.Endianness.BIG_ENDIAN), src__matcher__core_matchers.equals(1023));
      src__matcher__expect.expect(bd[dartx.getUint32](0, typed_data.Endianness.LITTLE_ENDIAN), src__matcher__core_matchers.equals(4278386688));
    }, VoidTodynamic()));
  };
  dart.fn(typed_arrays_dataview_test.main, VoidTodynamic());
  // Exports:
  exports.typed_arrays_dataview_test = typed_arrays_dataview_test;
});
